### DISCUSSION TITLE
Tests for the HA integration

### DISCUSSION THREAD

**Discussion by bdr99** [DurgNomis-drol/ha_toyota#13] [Ideas]
In order to create tests for the HA integration, we will need fixtures, or examples of the data returned by the methods in `mytoyota`. [Here](https://github.com/home-assistant/core/tree/dev/tests/fixtures/mazda) is an example from my Mazda integration. It looks like `MyT.gather_information` is the only method that the HA integration uses from `mytoyota`. @DurgNomis-drol Can you please provide a sample response for the `gather_information` method with any personal info censored?

Once we have the fixture(s), we can write the tests. The tests will create an invisible ("mock") instance of the Home Assistant application and initialize the Toyota integration, and then each test will verify that a specific part of the integration is working correctly. We will need to test the initialization of the integration, the config flow, and the entity classes (sensor, device tracker). By using the fixtures, we will be able to test the integration without actually connecting to Toyota servers.

Because the tests initialize an instance of HA, they will need to be run from within the `home-assistant/core` codebase. We can probably store them in this repo for now, but in order to run the tests, they will need to be copied to the proper location in `home-assistant/core`.

**Comment by DurgNomis-drol** [comment]
Thanks for all that information. This helps me a lot. I get my car on Friday and will look into it in the weekend 😄 I greatly appreciate that we can help each other.

I have changed somethings in the underlying `mytoyota` package, so i will have to update somethings first. But that should not take much time.

**Comment by edwinoonk** [comment]
Would you like to have access to my MyT account?

**Reply by bdr99** [reply]
Thank you, but I don't think that would be necessary, at least not for writing tests. All we need is sample responses to the methods in `mytoyota`, in order to create fixtures for the tests.

**Reply by DurgNomis-drol** [reply]
Not for writing these types of tests. 

But i made need your help with some other things soon. 😊

**Comment by edwinoonk** [comment]
Sure, I'd be happy to help

Op za 27 mrt. 2021 16:47 schreef Simon Hansen ***@***.***>:

**Comment by CM000n** [comment]
Hi @DurgNomis-drol , I recently switched jobs at work and was able to gain some experience in unit and integration testing myself.
Not for Home Assistant integrations but for Apache Spark data streams 😀 But if you still need help here, I would be happy to take a look.
Do you have something like a development branch that you are working on?

**Reply by DurgNomis-drol** [reply]
@CM000n I have been busy with another project for awhile now, though haven't forgot this. I would love some help with this as i have never really worked with unit testing or integration testing for HA. If you want, i can create a fork of home-assistant/core with this integration added so we can start implementing it? 😃 Though we can only add one platform per PR. So we will need to remove `device_tracker` for the initial PR.

**Reply by CM000n** [reply]
Sure, let's do it :-)

**Comment by DurgNomis-drol** [comment]
Hey guys, as you may have noticed, i have been spending a lot of time on this project the last month. Right now I have alot of things in my personal life, and because i am the only one right now maintaining this project (Though @CM000n are willing to help write tests). I have decided not to be the one to submit the PR for getting this into CORE. But if someone else is willing too do it they are more the welcome and i will make them a maintainer in this Repo. I will still be actively trying to fix bugs. I will how ever do some work on implementing statistics sensors as i feel this feature is more important and very useful. 

PR's are more the welcome! 

Sorry for not being upfront sooner 😃 Hopefully i will find more time in the coming months again.

Have a great day!
